soc: silabs_exx32: Add support for ARM ETM on gecko Series 2 SoCs

LeGrosFaisan opened the pull request:
This commit adds support for ARM ETM tracing feature on Silicon Labs
Gecko Series 2. It required some pins and clocks to be set.

Add TRACING_ETM as new tracing format in the subsys/tracing/Kconfig.

It was tested on the EFR32MG21 SoC using the `samples/synchronization` and Segger Ozone tool.

Signed-off-by: Steven Lemaire <steven.lemaire@zii.aero>

LeGrosFaisan committed:
soc: silabs_exx32: Add support for ARM ETM on gecko Series 2 SoCs

This commit adds support for ARM ETM tracing feature on Silicon Labs
Gecko Series 2. It required some pins and clocks to be set.

Add TRACING_ETM as new tracing format in the subsys/tracing/Kconfig.

Signed-off-by: Steven Lemaire <steven.lemaire@zii.aero>

chrta reviewed (dismissed):
Code looks good, i could not test this since i do not have a supported series 2 board.

tleksell-pe commented:
This is an interesting PR which highlight some aspects of software/hardware based tracing and board architecture features which needs to be dealt with:

1. ETM is a hardware based trace feature and should not be included with the software based tracing formats (choices) since these aren't mutually exclusive. That is, you should be able to run ETM in parallel with a software based trace format.
2. ETM is an optional hardware component and should be marked per board if it is supported (and use this to show/hide the enable ETM tracing option).

I would suggest the addition of a HAS_ARM_ETM (see HAS_SEGGER_RTT as an example) flag that boards can select together with a new choice/menu category for enabling it (hardware trace?).

LeGrosFaisan committed:
tracing: add a choice between hardware and software tracing

ETM is a hardware based trace feature, it does not requires to enable any
other tracing libraries to capture instructions' information.

Since it is a hardware feature, each ARM CPU/MCU series Kconfig must
explicitly say whether the hardware support it.

Signed-off-by: Steven Lemaire <steven.lemaire@zii.aero>

l-alfred commented:
@tleksell-pe I took the liberty to add the two points you mentioned in your comment as they are quite relevant. I thank you for that.
Please have a look at the second commit and tell me if it makes sense.

l-alfred commented:
I would like to re-open this PR if that is okay. Please let me know if the PR makes sense.
